In [1]:
from pandas import *
import numpy as np
from scipy.stats.stats import pearsonr
import itertools

from datacontainer import *
import sys
np.set_printoptions(threshold=sys.maxsize)
#%Read the file in%


## Load Data

Load data and header from a dataframe.

In [15]:
datamatrix = DataContainer('FINAL 102518 All Unique Metab KO.csv').dataMatrix
header = DataContainer('FINAL 102518 All Unique Metab KO.csv').header
# make  a dataframe
df=DataFrame(np.array(datamatrix))


## Correlation Analysis of Features
Find the pairwise correlations for the feature


In [16]:
correlations = {}
columns = df.columns.tolist()

for col_a, col_b in itertools.combinations(columns, 2):
    correlations[str(col_a) + '__' + str(col_b) +  "/(%s" % header[col_a] +"-%s)" % header[col_b]] = pearsonr(df.loc[:, col_a], df.loc[:, col_b])

result = DataFrame.from_dict(correlations, orient='index')
result.columns = ['PCC', 'p-value']

print(result.sort_values('PCC',ascending=False))


                                              PCC        p-value
7__24/(nof_Atoms-nof_Fragments)          1.000000   0.000000e+00
5__24/(molArea-nof_Fragments)            0.993586  1.246462e-130
5__7/(molArea-nof_Atoms)                 0.993586  1.246462e-130
1__7/(molVolume-nof_Atoms)               0.993554  1.745201e-130
1__24/(molVolume-nof_Fragments)          0.993554  1.745201e-130
1__5/(molVolume-molArea)                 0.991247  1.752824e-121
1__16/(molVolume-a_heavy)                0.982288  8.547405e-101
19__23/(C_R0-nof_RotB)                   0.977995   1.899211e-94
16__24/(a_heavy-nof_Fragments)           0.970694   4.309652e-86
7__16/(nof_Atoms-a_heavy)                0.970694   4.309652e-86
5__16/(molArea-a_heavy)                  0.968668   3.794397e-84
7__20/(nof_Atoms-C_sp3)                  0.964032   3.858290e-80
20__24/(C_sp3-nof_Fragments)             0.964032   3.858290e-80
1__20/(molVolume-C_sp3)                  0.938138   1.661728e-64
5__20/(molArea-C_sp3)    

`
Large Positive Correlations (>0.87):
                                              PCC        p-value
x7__24/(nof_Atoms-nof_Fragments)          1.000000   0.000000e+00 z
x5__7/(molArea-nof_Atoms)                 0.993586  1.246462e-130 z
x5__24/(molArea-nof_Fragments)            0.993586  1.246462e-130 z
x1__24/(molVolume-nof_Fragments)          0.993554  1.745201e-130 z
x1__7/(molVolume-nof_Atoms)               0.993554  1.745201e-130 z
x1__5/(molVolume-molArea)                 0.991247  1.752824e-121 z
x1__16/(molVolume-a_heavy)                0.982288  8.547405e-101 z
x19__23/(C_R0-nof_RotB)                   0.977995   1.899211e-94 z
x16__24/(a_heavy-nof_Fragments)           0.970694   4.309652e-86 z
x7__16/(nof_Atoms-a_heavy)                0.970694   4.309652e-86 z
x5__16/(molArea-a_heavy)                  0.968668   3.794397e-84 z
x7__20/(nof_Atoms-C_sp3)                  0.964032   3.858290e-80 z
x20__24/(C_sp3-nof_Fragments)             0.964032   3.858290e-80 z
x1__20/(molVolume-C_sp3)                  0.938138   1.661728e-64 z
x5__20/(molArea-C_sp3)                    0.936010   1.542450e-63 z
x13__16/(Complexity-a_heavy)              0.927815   4.219761e-60 z
x4__9/(molPSA-nof_HBA)                    0.912658   1.066100e-54 z
x8__17/(nof_Chirals-C_R2)                 0.897164   4.146746e-50 z
x16__20/(a_heavy-C_sp3)                   0.896385   6.741688e-50 z
x8__18/(nof_Chirals-C_R1)                 0.895264   1.348274e-49 z
x4__10/(molPSA-nof_HBD)                   0.880536   6.194669e-46 z
Default keep: nof rings, C_R0, PSA/area, complexity, SO3H, OH, no_Chirals
Try to keep: sp3, pos charge/area, mol Volume, C_R2
Keep: 1, 20, 18, 13, 4
Discard: 24, 7, 5, 16, 8, 23 [('molArea', 5), ('nof_Atoms', 7), ('nof_Chirals', 8), ('a_heavy', 16), ('nof_RotB', 23), ('nof_Fragments', 24)]
`


## Delete Redundant Columns


In [17]:
todel = [24, 7, 5, 16, 8, 23]
df.drop(df.columns[todel], axis=1,inplace=True)

In [18]:
print('features removed')
print([(header[x], x) for x in sorted(todel)])
print()
print('kept', [header[x]for x in sorted([24, 19, 4, 8])])




features removed
[('molArea', 5), ('nof_Atoms', 7), ('nof_Chirals', 8), ('a_heavy', 16), ('nof_RotB', 23), ('nof_Fragments', 24)]

kept ['molPSA', 'nof_Chirals', 'C_R0', 'nof_Fragments']
